### ISSUE TITLE
Schema directive context mutation is lost

### ISSUE THREAD

**Issue by alfaproject** [Urigo/graphql-modules#326]
I've just spent a great deal of time trying to figure out why my context changes were not being propagated to the original resolver in a schema directive and I'm not sure if it's inteded behaviour or not.

Schema directive code looks something like:
```ts
async function guard(source: unknown, args: Record<string, unknown>, context: Context, info: GraphQLResolveInfo) {
  // Fetch current user ACL context
  context.acl = 'whatever';
}

export class GuardDirective extends SchemaDirectiveVisitor {
  public visitFieldDefinition(field: GraphQLField<unknown, Context>): GraphQLField<unknown, Context> | void | null {
    const originalResolver = field.resolve || defaultFieldResolver;
    field.resolve = async (source, originalArgs, context, info) => {
      console.log('before guard context:', context.acl);
      await guard(source, originalArgs, context, info);
      console.log('after guard context:', context.acl);

      console.log('before original resolver context:', context.acl);
      await originalResolver.apply(field, [source, originalArgs, context, info]);
      console.log('after original resolver context:', context.acl);
    };
  }
}
```

The output of that in a guarded field looks like:
```
    console.log src/graphql/modules/guard.directive.ts:98
      before guard context: undefined

    console.log src/graphql/modules/guard.directive.ts:100
      after guard context: whatever

    console.log src/graphql/modules/guard.directive.ts:102
      before original resolver context: whatever

    console.log __tests__/graphql/modules/acl.spec.ts:19
      original resolver context: undefined

    console.log src/graphql/modules/guard.directive.ts:104
      after original resolver context: whatever
```

The context passed to the original resolver has `acl` set, but the original resolver method got a completely different context which doesn't have `acl` set.

As a workaround, I'm going to put my ACL context in the `GraphQLInfo` field since that one isn't touched, but that feels hacky.

**Comment by ardatan** [comment]
Why do you change the context using a directive? You can use `context` of `GraphQLModule`.

**Comment by alfaproject** [comment]
I can't do that because that's the global context. The only thing I do there is getting the user from the access token.

This is a guard directive which doubles as middleware and that's why I want to change the context that goes into the original resolver:
```
const { schema, context } = new GraphQLModule({
  imports: [aclModule],
  typeDefs: gql`
    type Query {
      guarded: String @guard
    }
  `,
  resolvers: {
    Query: {
      guarded: (source, args, ctx) => console.log('original resolver context:', ctx.acl),
    },
  },
});
```

That allows me to see whether or not the current user can query for the `guarded` field based on my attribute-based access control logic.

**Comment by ardatan** [comment]
GraphQL-Modules calls resolver functions with the context function of the module not the application. But we should figure this out. We will work on that!

**Comment by alfaproject** [comment]
Yeah, I saw that blog post but that's very basic for our needs. Our whole ACL directive is just `@guard` without any parameters. All the magic is done on the directive and that's why I need to pass the grant to the resolver. (The permissions are set by the users in the admin panel and they can set stuff like, a user with role X from organization Y can access resource `User` but can't see the `email` field for users in different organizations, etc)

I can't really share the business logic for all of that. What I can say is that the permissions are not hard-coded like in that blog post. Also, it's a very complex application with multi-tenancy and multi-layered roles so things get really complex when it comes to ACL.

My workaround, for now, was to mutate the `info` argument of the resolver, but I think using the `context` would make more sense.

Thank you for looking into that. (:

**Comment by argosystem-dev** [comment]
I have a similar problem. I used resolversComposition as a workaround for now..
the auth/acl is a bit less complicated, because I plan to set the required permission in the schema and check against the permissions in context (context.auth.permissions: string[])
with `query.graphql` schema in my user module I had
```
users(queryInput: QueryInput = {}): [User] @auth(permission: "user.listUsers")
user(id: Int!): User! @auth(permission: "user.readUser")
```

and I had to
`import { resolversComposition } from '@modules/auth/resolversComposition'`
in my `AppModule` (app.ts), because it did not work (was ignored) when it was imported in the auth module..but at least I had it working somehow... but then...

I upgraded to 0.7.0 and now it wont even compile:
```/usr/node/app/node_modules/apollo-server-core/src/ApolloServer.ts:254
throw new Error(errors.map(error => error.message).join('\n\n'));
^
 Error: Cannot use 'in' operator to search for 'subscribe' in undefined
at new ApolloServerBase (/usr/node/app/node_modules/apollo-server-core/src/ApolloServer.ts:254:15)
at new ApolloServer (/usr/node/app/node_modules/apollo-server-express/src/ApolloServer.ts:84:5)
at Object.<anonymous> (/usr/node/app/src/index.ts:16:16)
at Module._compile (internal/modules/cjs/loader.js:799:30)
at Module.m._compile (/usr/node/app/node_modules/ts-node/src/index.ts:439:23)
at Module._extensions..js (internal/modules/cjs/loader.js:810:10)
at Object.require.extensions.(anonymous function) [as .ts] (/usr/node/app/node_modules/ts-node/src/index.ts:442:12)
at Module.load (internal/modules/cjs/loader.js:666:32)
at tryModuleLoad (internal/modules/cjs/loader.js:606:12)
at Function.Module._load (internal/modules/cjs/loader.js:598:3)
at Function.Module.runMain (internal/modules/cjs/loader.js:862:12)
at Object.<anonymous> (/usr/node/app/node_modules/ts-node/src/bin.ts:154:12)
at Module._compile (internal/modules/cjs/loader.js:799:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:810:10)
at Module.load (internal/modules/cjs/loader.js:666:32)
at tryModuleLoad (internal/modules/cjs/loader.js:606:12)
```

So I think resolversComposition stopped working.. in 0.7.0

ps.
I don't know whats worse, hijacking an issue (if it's actually about a different bug...then it's even worse)
or opening a duplicate issue even if it's the same bug

**Comment by dancon** [comment]
I hava the same issue:

// user resolvers.ts

````
export default {
   Query: {
        getUserInfo (_, __, { user }) {
             // @tag0
             return user
        }
   }
}
```

// a middleware for authentication
```
const authenticated = (next) => (root, args, context, info) => {
     // some code
     // @tag1
     return next(root, args, context, info)
}
```

```
const module = new GraphQLModule({
     imports: [ userModule ],
     context () {
         // some code to get user
        const user = {...}
        return {
            user
       }
    },
    resolversComposition: {
         'Query.getUserInfo': [ authenticated ]
   }
})
```

I add some comments in the code snippets above `@tag0` and `@tag1`

The problem is I can get `user` from `context` at @tag1, but `user` is undefined at @tag0.

**Comment by ardatan** [comment]
This is not a problem, your composition belongs to the module that has context while resolver itself belongs to another module that don't have that context. I recommend you to move composition and context to `user` module.

**Comment by dancon** [comment]
@ardatan  got it, thx ~

**Comment by ardatan** [comment]
@alfaproject What do you pass as a context builder to your server config?

**Comment by alfaproject** [comment]
I'm no longer using `graphql-modules` because it increased the size of my bundled lambdas by more than 1mb, so I can't show any real code right now, but it used to be a small async function to decode a jwt and set the `userId` so that I could then fetch the ACL setup for the user.

**Comment by ardatan** [comment]
@alfaproject Sorry to hear that. 
Which bundler were you using, and what was the version when you stopped using graphql-modules?

**Comment by alfaproject** [comment]
I'm using webpack and I stopped using when this was released: https://github.com/Urigo/graphql-modules/issues/322

(because it didn't shake away TypeScript because it was (is?) a transitive dependency of `graphql-toolkit` / `graphql-tag-pluck`)

**Comment by ardatan** [comment]
I see but we fixed it since 0.6.2. And you can see the comparison and current bundle size
from here;
https://bundlephobia.com/result?p=@graphql-modules/core@0.7.5
![](https://files.slack.com/files-pri/TAYJ1FSUA-FKY72EGSG/image.png)
![image](https://user-images.githubusercontent.com/20847995/60305885-29b69680-990c-11e9-9aef-2b0ac383ec86.png)

**Comment by alfaproject** [comment]
That's much better but I think some of your transitive dependencies use `commonjs` like the ones that import TypeScript which is more than 50% of the impact of adding this module.

I'd like to try this again and I'm happy you guys are aware of size tracking now, which is more than I can say about other authors.

If I try again, I'll create specific issues or PRs to help you get a minimal bundle with webpack. (:
